server/routes/middleware.js

```diff
@@ -0,0 +1,38 @@
+const jwt = require("jsonwebtoken");
+
+const verifyToken = (req, res, next) => {
+    // If no authorization header return unauthorized message.
+    if (!req.header("Authorization"))
+        return res.status(401).json({
+            success: false,
+            msg: "Unauthorized",
+        });
+
+    // Check for the authorization bearer header.
+    if (req.header("Authorization") && req.header("Authorization").split(" ")[0] === "Bearer") {
+        // Gets jwt token from the header.
+        const token = req.header("Authorization").split(" ")[1];
+
+        try {
+            // Verifies jwt token and returns the decoded user information.
+            // Throws an error if jwt is invalid (tampered with or encoded with a different secret).
+            const userInfo = jwt.verify(token, process.env.TOKEN_SECRET);
+
+            // Add user info to the request object and continue to next middleware
+            req.user = userInfo;
```

Comment:
Rather than adding the userInfo to the request should we just say that if userInfo.id doesn't match req.params.userId return an error message? This way the verification takes place here automatically rather than having to do the comparison in each function that needs to verify that the requesting user's Id matches the 1 in the url

Reply:
The token verification is going to be done on every request except for the login requests, therefore on some request you may not have the req.params.userId (eg. POST request).

Also we are not verifying the user ID when we verify the token. We verify that the user logging in is a valid user in the db before creating and issuing the token. On every request when the token is sent, we are trusting that the token is valid once it has been created using our token secret.

The updated express-jwt verification adds the user to the req object by default.

Reply:
Ok. Then we can just write a function in a util file in the controllers dir to do the comparison and take appropriate actions if the ids don't match (to reduce code duplication everywhere we want to compare the ids).

Reply:
I don't think we need to do the comparison of the id's. Once the token is verified then the user is authorized to make the request. No need to compare the ids after verification. Once verified, the request can be handled.

Reply:
The comparison was to make sure the requesting user is editing their own info. For example we don't want user A to somehow access the createConversation or a potential future updateUser route using user B's id in the url.

Reply:
True. In that case the comparison could done in a util file and passed as middleware to the requests that need it.